File: code/python/04-rag-pipeline/embedding_visualizer.py
"""Embedding visualization utilities for building semantic intuition.

Provides text similarity matrices, outlier detection, and K-means clustering
over embedding vectors — all without requiring matplotlib or a display.

See: docs/03-memory-and-retrieval/02-embeddings-and-vectors.md
"""

from __future__ import annotations

import math
from typing import Optional

import numpy as np

from embedding_generator import EmbeddingGenerator, EmbeddingComparator


class EmbeddingVisualizer:
    """Visualize embeddings to understand semantic relationships.

    Uses the provided :class:`EmbeddingGenerator` for all embedding calls.
    All output is plain text so it works in any terminal or log stream.

    Args:
        generator: Pre-configured :class:`EmbeddingGenerator` to use.
    """

    def __init__(self, generator: EmbeddingGenerator) -> None:
        self.generator = generator
        self._comparator = EmbeddingComparator()

    # ------------------------------------------------------------------
    # Similarity matrix
    # ------------------------------------------------------------------

    def compare_texts(
        self,
        texts: list[str],
        labels: Optional[list[str]] = None,
    ) -> str:
        """Generate a similarity matrix as a formatted Unicode table.

        Args:
            texts: Texts to compare pairwise.
            labels: Optional display labels. Defaults to "Text 1", "Text 2", …

        Returns:
            Multi-line string ready to print.

        Example output::

            ┌──────────────┬────────┬─────────┬────────┐
            │              │ Text 1 │  Text 2 │ Text 3 │
            ├──────────────┼────────┼─────────┼────────┤
            │ Text 1       │  1.00  │   0.85  │  0.12  │
            │ Text 2       │  0.85  │   1.00  │  0.15  │
            │ Text 3       │  0.12  │   0.15  │  1.00  │
            └──────────────┴────────┴─────────┴────────┘
        """
        if labels is None:
            labels = [f"Text {i + 1}" for i in range(len(texts))]
        if len(labels) != len(texts):
            raise ValueError("len(labels) must equal len(texts)")

        embeddings = self.generator.embed_batch(texts)
        n = len(texts)
        scores = [
            [
                self._comparator.cosine_similarity(embeddings[i], embeddings[j])
                for j in range(n)
            ]
            for i in range(n)
        ]

        col_w = 8
        label_w = max(len(lbl) for lbl in labels) + 2
        h_line = "─" * label_w
        col_lines = ["─" * col_w for _ in labels]

        top = "┌" + h_line + "┬" + "┬".join(col_lines) + "┐"
        header_cells = "│" + " " * label_w + "│" + "│".join(
            f" {lbl:^{col_w - 1}}" for lbl in labels
        ) + "│"
        divider = "├" + h_line + "┼" + "┼".join(col_lines) + "┤"
        bottom = "└" + h_line + "┴" + "┴".join(col_lines) + "┘"

        data_rows = []
        for i in range(n):
            cells = "│".join(f" {scores[i][j]:^{col_w - 1}.2f}" for j in range(n))
            data_rows.append("│" + f" {labels[i]:<{label_w - 1}}" + "│" + cells + "│")

        return "\n".join([top, header_cells, divider] + data_rows + [bottom])

    # ------------------------------------------------------------------
    # Outlier detection
    # ------------------------------------------------------------------

    def find_outliers(
        self,
        texts: list[str],
        labels: Optional[list[str]] = None,
        threshold_percentile: float = 25.0,
    ) -> list[str]:
        """Find texts that are semantically different from all others.

        A text is considered an outlier if its average similarity to all
        other texts falls below the *threshold_percentile* of the pairwise
        distribution.

        Args:
            texts: Texts to analyse.
            labels: Display labels (defaults to the texts themselves).
            threshold_percentile: Percentile cutoff for the average similarity.

        Returns:
            List of labels (or texts, if no labels are provided) identified
            as outliers.
        """
        if labels is None:
            labels = texts

        embeddings = self.generator.embed_batch(texts)
        n = len(texts)
        avg_sims = []
        for i in range(n):
            sims = [
                self._comparator.cosine_similarity(embeddings[i], embeddings[j])
                for j in range(n)
                if j != i
            ]
            avg_sims.append(sum(sims) / len(sims) if sims else 1.0)

        cutoff = float(np.percentile(avg_sims, threshold_percentile))
        return [labels[i] for i in range(n) if avg_sims[i] < cutoff]

    # ------------------------------------------------------------------
    # K-means clustering
    # ------------------------------------------------------------------

    def cluster(
        self,
        texts: list[str],
        n_clusters: int = 3,
        max_iterations: int = 100,
        random_seed: int = 42,
    ) -> dict[int, list[str]]:
        """Group texts into semantic clusters using K-means on their embeddings.

        This is a plain NumPy implementation — no scikit-learn required.

        Args:
            texts: Texts to cluster.
            n_clusters: Number of clusters to produce.
            max_iterations: Maximum K-means iterations.
            random_seed: Seed for reproducibility.

        Returns:
            Dict mapping cluster index → list of texts in that cluster.
        """
        embeddings = self.generator.embed_batch(texts)
        matrix = np.array(embeddings, dtype=np.float64)  # shape (n, dim)
        n = len(texts)

        rng = np.random.default_rng(random_seed)
        centroid_indices = rng.choice(n, size=n_clusters, replace=False)
        centroids = matrix[centroid_indices].copy()

        assignments = np.zeros(n, dtype=int)
        for _ in range(max_iterations):
            # Assign each text to the nearest centroid.
            new_assignments = np.array(
                [_nearest_centroid(matrix[i], centroids) for i in range(n)],
                dtype=int,
            )
            if np.array_equal(new_assignments, assignments):
                break
            assignments = new_assignments

            # Recompute centroids.
            for k in range(n_clusters):
                members = matrix[assignments == k]
                if len(members) > 0:
                    centroids[k] = members.mean(axis=0)

        result: dict[int, list[str]] = {k: [] for k in range(n_clusters)}
        for i, text in enumerate(texts):
            result[assignments[i]].append(text)
        return result

    # ------------------------------------------------------------------
    # Built-in demo
    # ------------------------------------------------------------------

    def demo_semantic_categories(self) -> None:
        """Show clear semantic clustering across weather, technology, and food.

        Demonstrates:
        - Within-category similarity > 0.65
        - Cross-category similarity < 0.30
        - K-means clustering correctly groups all 9 texts into 3 clusters
        """
        categories = {
            "weather": [
                "It's raining outside",
                "Sunny with a high of 75",
                "Snow expected tomorrow",
            ],
            "technology": [
                "Python 3.12 released",
                "New GPU architecture announced",
                "API deprecation notice",
            ],
            "food": [
                "Best pizza in New York",
                "How to make sourdough bread",
                "Restaurant review: French cuisine",
            ],
        }

        all_texts = [t for texts in categories.values() for t in texts]
        all_labels = list(categories)
        label_per_text = [
            lbl for lbl, texts in categories.items() for _ in texts
        ]

        print("=== Similarity matrix (9 texts × 9 texts) ===")
        short_labels = [
            f"{lbl[:4]}#{i + 1}"
            for lbl, texts in categories.items()
            for i in range(len(texts))
        ]
        print(self.compare_texts(all_texts, labels=short_labels))

        embeddings = self.generator.embed_batch(all_texts)

        print("\n=== Within-category similarities (expect > 0.65) ===")
        for cat_name, texts in categories.items():
            cat_indices = [i for i, lbl in enumerate(label_per_text) if lbl == cat_name]
            pairs = [
                (i, j)
                for i in cat_indices
                for j in cat_indices
                if i < j
            ]
            for i, j in pairs:
                sim = self._comparator.cosine_similarity(embeddings[i], embeddings[j])
                status = "✓" if sim > 0.65 else "✗"
                print(f"  {status} [{cat_name}]  {sim:.3f}")

        print("\n=== Cross-category similarities (expect < 0.30) ===")
        cat_names = list(categories)
        for ci in range(len(cat_names)):
            for cj in range(ci + 1, len(cat_names)):
                cat_a, cat_b = cat_names[ci], cat_names[cj]
                indices_a = [i for i, lbl in enumerate(label_per_text) if lbl == cat_a]
                indices_b = [i for i, lbl in enumerate(label_per_text) if lbl == cat_b]
                sims = [
                    self._comparator.cosine_similarity(embeddings[ia], embeddings[ib])
                    for ia in indices_a
                    for ib in indices_b
                ]
                avg = sum(sims) / len(sims)
                status = "✓" if avg < 0.30 else "✗"
                print(f"  {status} [{cat_a} vs {cat_b}]  avg={avg:.3f}")

        print("\n=== K-means clustering (k=3) ===")
        clusters = self.cluster(all_texts, n_clusters=3)
        for cluster_id, members in clusters.items():
            print(f"  Cluster {cluster_id}:")
            for text in members:
                print(f"    - {text}")


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _nearest_centroid(vector: np.ndarray, centroids: np.ndarray) -> int:
    """Return the index of the centroid closest to *vector* (cosine distance)."""
    sims = centroids @ vector / (
        np.linalg.norm(centroids, axis=1) * np.linalg.norm(vector) + 1e-10
    )
    return int(np.argmax(sims))


# ---------------------------------------------------------------------------
# Demo entry point
# ---------------------------------------------------------------------------


def main() -> None:
    """Run the built-in semantic categories demo."""
    gen = EmbeddingGenerator(model="text-embedding-3-small")
    viz = EmbeddingVisualizer(gen)
    viz.demo_semantic_categories()

    # --- Outlier detection ---
    texts = [
        "The weather is sunny today.",
        "It will rain tomorrow afternoon.",
        "Clear skies expected all week.",
        "The quarterly earnings beat expectations.",  # outlier in weather context
    ]
    print("\n=== Outlier detection ===")
    outliers = viz.find_outliers(texts)
    print(f"  Outliers: {outliers}")


if __name__ == "__main__":
    main()

File: code/go/04-rag-pipeline/embedding_visualizer.go
// embedding_visualizer.go — Embedding visualization utilities.
//
// Provides text similarity matrices, outlier detection, and K-means clustering
// over embedding vectors — all without requiring external plotting libraries.
//
// See: docs/03-memory-and-retrieval/02-embeddings-and-vectors.md
package ragpipeline

import (
	"context"
	"fmt"
	"math"
	"math/rand"
	"sort"
	"strings"
)

// ---------------------------------------------------------------------------
// EmbeddingVisualizer
// ---------------------------------------------------------------------------

// EmbeddingVisualizer generates text representations of embedding relationships.
type EmbeddingVisualizer struct {
	Generator  *EmbeddingGenerator
	Comparator *EmbeddingComparator
}

// NewEmbeddingVisualizer creates an EmbeddingVisualizer.
func NewEmbeddingVisualizer(generator *EmbeddingGenerator) *EmbeddingVisualizer {
	return &EmbeddingVisualizer{
		Generator:  generator,
		Comparator: &EmbeddingComparator{},
	}
}

// ---------------------------------------------------------------------------
// Similarity matrix
// ---------------------------------------------------------------------------

// CompareTexts generates a similarity matrix as a formatted Unicode table.
func (ev *EmbeddingVisualizer) CompareTexts(ctx context.Context, texts []string, labels []string) (string, error) {
	if labels == nil {
		labels = make([]string, len(texts))
		for i := range texts {
			labels[i] = fmt.Sprintf("Text %d", i+1)
		}
	}
	if len(labels) != len(texts) {
		return "", fmt.Errorf("len(labels) must equal len(texts)")
	}

	embeddings, err := ev.Generator.EmbedBatch(ctx, texts)
	if err != nil {
		return "", err
	}
	n := len(texts)

	// Compute pairwise similarities.
	scores := make([][]float64, n)
	for i := range scores {
		scores[i] = make([]float64, n)
		for j := range scores[i] {
			sim, _ := ev.Comparator.CosineSimilarity(embeddings[i], embeddings[j])
			scores[i][j] = sim
		}
	}

	// Format as table.
	colW := 8
	labelW := 2
	for _, lbl := range labels {
		if len(lbl)+2 > labelW {
			labelW = len(lbl) + 2
		}
	}
	hLine := strings.Repeat("─", labelW)
	colLines := make([]string, n)
	for i := range colLines {
		colLines[i] = strings.Repeat("─", colW)
	}

	colsStr := strings.Join(colLines, "┬")
	top := "┌" + hLine + "┬" + colsStr + "┐"
	var headerCells strings.Builder
	headerCells.WriteString("│" + strings.Repeat(" ", labelW))
	for _, lbl := range labels {
		cell := fmt.Sprintf(" %-*s", colW-1, lbl)
		if len(cell) > colW {
			cell = cell[:colW]
		}
		headerCells.WriteString("│" + cell)
	}
	headerCells.WriteString("│")
	divider := "├" + hLine + "┼" + strings.Join(colLines, "┼") + "┤"
	bottom := "└" + hLine + "┴" + strings.Join(colLines, "┴") + "┘"

	var rows []string
	for i := 0; i < n; i++ {
		lbl := fmt.Sprintf(" %-*s", labelW-1, labels[i])
		var row strings.Builder
		row.WriteString("│" + lbl)
		for j := 0; j < n; j++ {
			cell := fmt.Sprintf(" %*.2f", colW-1, scores[i][j])
			row.WriteString("│" + cell)
		}
		row.WriteString("│")
		rows = append(rows, row.String())
	}

	parts := []string{top, headerCells.String(), divider}
	parts = append(parts, rows...)
	parts = append(parts, bottom)
	return strings.Join(parts, "\n"), nil
}

// ---------------------------------------------------------------------------
// Outlier detection
// ---------------------------------------------------------------------------

// FindOutliers finds texts that are semantically different from all others.
// A text is an outlier if its average similarity to all others is below
// the threshold_percentile of the pairwise distribution.
func (ev *EmbeddingVisualizer) FindOutliers(ctx context.Context, texts []string, labels []string, thresholdPercentile float64) ([]string, error) {
	if labels == nil {
		labels = texts
	}

	embeddings, err := ev.Generator.EmbedBatch(ctx, texts)
	if err != nil {
		return nil, err
	}
	n := len(texts)
	avgSims := make([]float64, n)
	for i := range texts {
		total := 0.0
		for j := range texts {
			if i != j {
				sim, _ := ev.Comparator.CosineSimilarity(embeddings[i], embeddings[j])
				total += sim
			}
		}
		if n > 1 {
			avgSims[i] = total / float64(n-1)
		} else {
			avgSims[i] = 1.0
		}
	}

	// Compute percentile cutoff.
	sorted := make([]float64, n)
	copy(sorted, avgSims)
	sort.Float64s(sorted)
	cutoffIdx := int(math.Floor(thresholdPercentile / 100.0 * float64(n)))
	if cutoffIdx >= n {
		cutoffIdx = n - 1
	}
	cutoff := sorted[cutoffIdx]

	var outliers []string
	for i, sim := range avgSims {
		if sim < cutoff {
			outliers = append(outliers, labels[i])
		}
	}
	return outliers, nil
}

// ---------------------------------------------------------------------------
// K-means clustering
// ---------------------------------------------------------------------------

// Cluster groups texts into semantic clusters using K-means on their embeddings.
// Returns a map of cluster index → list of texts in that cluster.
func (ev *EmbeddingVisualizer) Cluster(ctx context.Context, texts []string, nClusters, maxIterations int, randomSeed int64) (map[int][]string, error) {
	if nClusters > len(texts) {
		nClusters = len(texts)
	}
	embeddings, err := ev.Generator.EmbedBatch(ctx, texts)
	if err != nil {
		return nil, err
	}
	n := len(texts)
	dim := len(embeddings[0])

	rng := rand.New(rand.NewSource(randomSeed))

	// Choose k random initial centroids.
	centroidIdxs := rng.Perm(n)[:nClusters]
	centroids := make([][]float64, nClusters)
	for i, idx := range centroidIdxs {
		centroids[i] = make([]float64, dim)
		copy(centroids[i], embeddings[idx])
	}

	assignments := make([]int, n)

	for iter := 0; iter < maxIterations; iter++ {
		// Assign each text to the nearest centroid (by cosine similarity).
		changed := false
		for i := range texts {
			best := 0
			bestSim := cosineSim(embeddings[i], centroids[0])
			for k := 1; k < nClusters; k++ {
				sim := cosineSim(embeddings[i], centroids[k])
				if sim > bestSim {
					bestSim = sim
					best = k
				}
			}
			if assignments[i] != best {
				assignments[i] = best
				changed = true
			}
		}
		if !changed {
			break
		}

		// Recompute centroids.
		for k := 0; k < nClusters; k++ {
			newCentroid := make([]float64, dim)
			count := 0
			for i, a := range assignments {
				if a == k {
					for d := range newCentroid {
						newCentroid[d] += embeddings[i][d]
					}
					count++
				}
			}
			if count > 0 {
				for d := range newCentroid {
					newCentroid[d] /= float64(count)
				}
				centroids[k] = newCentroid
			}
		}
	}

	result := make(map[int][]string, nClusters)
	for k := 0; k < nClusters; k++ {
		result[k] = nil
	}
	for i, text := range texts {
		result[assignments[i]] = append(result[assignments[i]], text)
	}
	return result, nil
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

// DemoSemanticCategories shows clear semantic clustering across categories.
func (ev *EmbeddingVisualizer) DemoSemanticCategories(ctx context.Context) {
	categories := map[string][]string{
		"weather": {
			"It's raining outside",
			"Sunny with a high of 75",
			"Snow expected tomorrow",
		},
		"technology": {
			"Python 3.12 released",
			"New GPU architecture announced",
			"API deprecation notice",
		},
		"food": {
			"Best pizza in New York",
			"How to make sourdough bread",
			"Restaurant review: French cuisine",
		},
	}

	var allTexts []string
	var shortLabels []string
	catNames := []string{"weather", "technology", "food"}
	for _, cat := range catNames {
		texts := categories[cat]
		for i, t := range texts {
			allTexts = append(allTexts, t)
			shortLabels = append(shortLabels, fmt.Sprintf("%s#%d", cat[:4], i+1))
		}
	}

	fmt.Println("=== Similarity matrix (9 texts × 9 texts) ===")
	table, err := ev.CompareTexts(ctx, allTexts, shortLabels)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	fmt.Println(table)

	fmt.Println("\n=== K-means clustering (k=3) ===")
	clusters, err := ev.Cluster(ctx, allTexts, 3, 100, 42)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	for clusterID, members := range clusters {
		fmt.Printf("  Cluster %d:\n", clusterID)
		for _, text := range members {
			fmt.Printf("    - %s\n", text)
		}
	}

	fmt.Println("\n=== Outlier detection ===")
	weatherPlusOutlier := []string{
		"The weather is sunny today.",
		"It will rain tomorrow afternoon.",
		"Clear skies expected all week.",
		"The quarterly earnings beat expectations.",
	}
	outliers, err := ev.FindOutliers(ctx, weatherPlusOutlier, nil, 25.0)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	fmt.Printf("  Outliers: %v\n", outliers)
}

// RunEmbeddingVisualizer demonstrates the embedding visualizer.
func RunEmbeddingVisualizer() {
	gen := NewEmbeddingGenerator("text-embedding-3-small", 0)
	viz := NewEmbeddingVisualizer(gen)
	viz.DemoSemanticCategories(context.Background())
}
